component/discovery/kubernetes_crds/kubernetes_crds.go

```diff
@@ -0,0 +1,272 @@
+	BearerTokenFile string `json:"bearerTokenFile,omitempty"`
+	// TLS Config to use for accessing apiserver.
+	TLSConfig commonConfig.TLSConfig `json:"tlsConfig,omitempty"`
+	// Authorization section for accessing apiserver
+	Authorization commonConfig.Authorization `json:"authorization,omitempty"`
+}
+
+func (c *Config) restConfig() (*rest.Config, error) {
+	if c.KubeConfig != "" {
+		return clientcmd.BuildConfigFromFlags("", c.KubeConfig)
+	}
+	if c.ApiServerConfig == nil {
+		return rest.InClusterConfig()
+	}
+	// TODO
+	log.Fatal("Convert apiserverconfig directly")
+	return nil, nil
+}
+
+type Component struct {
+	opts      component.Options
+	discovery *discovery.Manager
+	scraper   *scrape.Manager
+
+	config           *Config
+	cg               configGenerator
+	discoveryConfigs map[string]discovery.Configs
+	scrapeConfigs    map[string]*config.ScrapeConfig
+
+	onUpdate chan struct{}
+	mut      sync.Mutex
+}
+
+func New(o component.Options, args component.Arguments) (*Component, error) {
+
+	c := &Component{
+		opts:             o,
+		discoveryConfigs: map[string]discovery.Configs{},
+		scrapeConfigs:    map[string]*config.ScrapeConfig{},
+		onUpdate:         make(chan struct{}, 1),
+	}
+	return c, c.Update(args)
+}
+
+func (c *Component) apply() {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	c.discovery.ApplyConfig(c.discoveryConfigs)
+	scs := []*config.ScrapeConfig{}
+	for _, sc := range c.scrapeConfigs {
+		scs = append(scs, sc)
+	}
+	err := c.scraper.ApplyConfig(&config.Config{
+		ScrapeConfigs: scs,
+	})
+	if err != nil {
+		level.Error(c.opts.Logger).Log("msg", "error applying scrape configs", "err", err)
+	}
+	level.Debug(c.opts.Logger).Log("msg", "scrape config was updated")
+}
+
+func (c *Component) clearConfigs(kind string, ns string, name string) {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	prefix := fmt.Sprintf("%s/%s/%s", kind, ns, name)
+	for k := range c.discoveryConfigs {
+		if strings.HasPrefix(k, prefix) {
+			delete(c.discoveryConfigs, k)
+			delete(c.scrapeConfigs, k)
+		}
+	}
+}
+
+func (c *Component) addConfig(pm *v1.PodMonitor) {
+	for i, ep := range pm.Spec.PodMetricsEndpoints {
+		pmc := c.cg.generatePodMonitorConfig(pm, ep, i)
+		c.mut.Lock()
+		c.discoveryConfigs[pmc.JobName] = pmc.ServiceDiscoveryConfigs
+		c.scrapeConfigs[pmc.JobName] = pmc
+		c.mut.Unlock()
+	}
+	c.apply()
+}
+
+func (c *Component) OnAddPodMonitor(obj interface{}) {
+	pm := obj.(*v1.PodMonitor)
+	level.Info(c.opts.Logger).Log("msg", "found pod monitor", "name", pm.Name)
+	c.addConfig(pm)
+}
+
+func (c *Component) OnUpdatePodMonitor(oldObj, newObj interface{}) {
+	pm := oldObj.(*v1.PodMonitor)
+	level.Info(c.opts.Logger).Log("msg", "found pod monitor", "name", pm.Name)
+	c.clearConfigs("podMonitor", pm.Namespace, pm.Name)
+	c.addConfig(newObj.(*v1.PodMonitor))
+}
+func (c *Component) OnDeletePodMonitor(obj interface{}) {
+	pm := obj.(*v1.PodMonitor)
+	c.clearConfigs("podMonitor", pm.Namespace, pm.Name)
+	c.apply()
+}
```

Comment:
These three methods can be unexported; they don't need to be a part of the public API.